From 73dfc12c73748d30e0998bab6755385dadd78650 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20W=C3=B6rndl-Aichriedler?= <bwa@xdevelop.at>
Date: Sat, 15 May 2010 16:34:05 +0200
Subject: [PATCH 13/48] Add power-off support for the TWL4030 companion

This patch adds support for the power-off on shutdown feature of the TWL4030
---
 drivers/mfd/Kconfig    |    6 ++++++
 drivers/mfd/twl-core.c |   40 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 46 insertions(+)

diff --git a/drivers/mfd/Kconfig b/drivers/mfd/Kconfig
index f1391c2..ed66cc3 100644
--- a/drivers/mfd/Kconfig
+++ b/drivers/mfd/Kconfig
@@ -240,6 +240,12 @@ config MFD_TWL4030_AUDIO
 	select MFD_CORE
 	default n
 
+config TWL4030_POWEROFF
+	bool "TWL4030 Allow power-off on shutdown"
+	depends on TWL4030_CORE
+	help
+	  Enables the CPU to power-off the system on shutdown
+
 config TWL6030_PWM
 	tristate "TWL6030 PWM (Pulse Width Modulator) Support"
 	depends on TWL4030_CORE
diff --git a/drivers/mfd/twl-core.c b/drivers/mfd/twl-core.c
index 61e70cf..73e29fd 100644
--- a/drivers/mfd/twl-core.c
+++ b/drivers/mfd/twl-core.c
@@ -123,6 +123,12 @@
 #define twl_has_bci()	false
 #endif
 
+#if defined (CONFIG_TWL4030_POWEROFF)
+#define twl_has_poweroff()	true
+#else
+#define twl_has_poweroff()	false
+#endif
+
 /* Triton Core internal information (BEGIN) */
 
 /* Last - for index max*/
@@ -226,6 +232,10 @@
 #define TWL5031			BIT(2)  /* twl5031 has different registers */
 #define TWL6030_CLASS		BIT(3)	/* TWL6030 class */
 
+/* for pm_power_off */
+#define PWR_P1_SW_EVENTS	0x10
+#define PWR_DEVOFF		(1 << 0)
+
 /*----------------------------------------------------------------------*/
 
 /* is driver active, bound to a chip? */
@@ -1176,6 +1186,30 @@ static int twl_remove(struct i2c_client *client)
 	return 0;
 }
 
+static void twl_poweroff(void)
+{
+	int err;
+	u8 val;
+
+	err = twl_i2c_read_u8(TWL4030_MODULE_PM_MASTER, &val,
+				  PWR_P1_SW_EVENTS);
+	if (err) {
+		pr_err("%s: i2c error %d while reading TWL4030"
+			"PM_MASTER P1_SW_EVENTS\n",
+			DRIVER_NAME, err);
+		return;
+	}
+
+	val |= PWR_DEVOFF;
+
+	err = twl_i2c_write_u8(TWL4030_MODULE_PM_MASTER, val,
+				   PWR_P1_SW_EVENTS);
+	if (err)
+		pr_err("%s: i2c error %d while writing TWL4030"
+			"PM_MASTER P1_SW_EVENTS\n",
+			DRIVER_NAME, err);
+}
+
 /* NOTE:  this driver only handles a single twl4030/tps659x0 chip */
 static int __devinit
 twl_probe(struct i2c_client *client, const struct i2c_device_id *id)
@@ -1270,6 +1304,12 @@ twl_probe(struct i2c_client *client, const struct i2c_device_id *id)
 		twl_i2c_write_u8(TWL4030_MODULE_INTBR, temp, REG_GPPUPDCTR1);
 	}
 
+	if(twl_has_poweroff())
+	{
+		/* initialize pm_power_off routine */
+		pm_power_off = twl_poweroff;
+	}
+
 	status = add_children(pdata, id->driver_data);
 fail:
 	if (status < 0)
-- 
1.7.10.4

